crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1653,48 +1653,71 @@ impl<'db> TypeInferenceBuilder<'db> {
+                ast::FStringPart::Literal(literal) => {
+                    concatenated.push_str(&literal.value);
                 }
                 ast::FStringPart::FString(fstring) => {
-                    let ast::FString {
-                        range: _,
-                        elements,
-                        flags: _,
-                    } = fstring;
-                    for element in elements {
-                        self.infer_fstring_element(element);
+                    for element in &fstring.elements {
+                        match element {
+                            ast::FStringElement::Expression(expression) => {
+                                let ast::FStringExpressionElement {
+                                    range: _,
+                                    expression,
+                                    debug_text: _,
+                                    conversion,
+                                    format_spec,
+                                } = expression;
+                                // Always infer sub-expressions, even if we've figured out the type
+                                let ty = self.infer_expression(expression);
+                                if !done {
+                                    // TODO: handle format specifiers by calling a method
+                                    // (`Type::format`?) that handles the `__format__` method.
+                                    // Conversion flags should be handled before calling
+                                    // `__format__`.
+                                    // https://docs.python.org/3/library/string.html#format-string-syntax
+                                    if !conversion.is_none() || format_spec.is_some() {
+                                        has_expression = true;
+                                        done = true;
+                                    } else {
+                                        if let Some(Type::StringLiteral(literal)) = ty.str(self.db)
+                                        {
+                                            concatenated.push_str(literal.value(self.db));
+                                        } else {
+                                            has_expression = true;
+                                            done = true;
+                                        }
+                                    }
+                                }
+                            }
+                            ast::FStringElement::Literal(literal) => {
+                                if !done {
+                                    concatenated.push_str(&literal.value);
+                                }
+                            }
+                        }
                     }
                 }
             }
+            if concatenated.len() > Self::MAX_STRING_LITERAL_SIZE {
+                done = true;
+            }
         }
```

Comment:
Checking the string length only after appending to `concatenated` has the downside that the implementation copies over very long strings only to then decide that this was unnecessary

I recommend to eagerly check the string length before appending

```rust
struct Collector {
	concatenated: Option<String>,
	expression: bool,
}

impl Collector {
	fn new() -> Self {
		Self { concatenated: Some(String::new()), expression: false }
	}

	fn push_str(&mut self, literal: &str) {
		self.concatenated = self.concatenated.as_mut().and_then(|concatenated| {
			if concatenated.len().saturating_add(literal.len()) <= MAX_STRING_LITERAL_SIZE {
				concatenated.push_str(literal);
				Some(concatenated)
			} else {
				None
		})
	}

  fn add_expression(&mut self) {
		self.concatenated = None;
		self.expression = true;
	}

  fn ty(self, db: &dyn Db) -> Type {
		if self.expression {
			builtins_symbol_ty(self.db, "str").to_instance(db)
		} else if let Some(concatenated) = self.concatenated { 
			Type::StringLiteral(StringLiteralType::new(db, concatenated.into_boxed_str()))
		} else {
			Type::LiteralString
		}
	} 
}
```

Reply:
Thank you for this suggestion!